File: 프로그래머스/lv3/49191. 순위/순위.js
function solution(n, results) {
    const result = Array.from({length: n}, () => new Array(n).fill(false));

    for(const [win, lose] of results){
        result[win-1][lose-1] = true;
    }
    
     for (let k = 0; k < n; k++) {
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                if (result[i][k] && result[k][j]) {
                    result[i][j] = true;
                }
            }
        }
    }

    let answer = 0;
    for (let i = 0; i < n; i++) {
        let count = 0;
        for (let j = 0; j < n; j++) {
            if (result[i][j] || result[j][i]) {
                count++;
            }
        }
 
        if (count === n - 1) {
            answer++;
        }
    }

    return answer;
}

File: 프로그래머스/3/49191. 순위/순위.py
from collections import deque

def solution(n, results):
    answer = 0
    
    win = [[] for _ in range(n+1)]
    lose = [[] for _ in range(n+1)]
    
    for a,b in results:
        win[a].append(b)
        lose[b].append(a)
    
    def bfs(v,li):
        cnt = 0
        visited = [False] * (n+1)
        visited[v] = True
        q = deque([v])
    
        while q:
            nv = q.popleft()
        
            for i in li[nv]:
                if not visited[i]:
                    visited[i] = True
                    q.append(i)
                    cnt+= 1
    
        return cnt
    
    
    for i in range(1,n+1):
        win_cnt = bfs(i,win)
        lose_cnt = bfs(i,lose)
    
        if win_cnt + lose_cnt == n-1:
            answer += 1
    
    
    return answer
            